doc: add a spec for the change to formatted copying (#5212)
microsoft/terminal: doc/specs/#4191 - Formatted Copy/spec.md

---
author: Kayla Cinnamon @cinnamon-msft
created on: <2020-04-01>
last updated: <2020-04-01>
issue id: <#4191>
---

# Formatted Copy

## Abstract

When copying text, the Terminal should provide the option of including formatting. Not all apps that receive text allow for picking which format you want when pasting. The default should be to only copy plain text, based on the response from this poll on Twitter.

![Twitter poll](twitter-poll.png)

## Solution Design

Solutions for the right click behavior and user settings are described below.

### Right click behavior

Be default, right clicking to copy would only copy the plain text. If the user would like to copy the formatting, they can hold `alt` and right click (this behavior is the same in conhost).

### Settings option 1 - global setting

We could have a global setting that adds the formatting when the user copies. When set, this would change the right click behavior.

### Settings option 2 - key binding argument

> We could add an argument to the `copy` key binding to allow for formatted copying when the user chooses to do so.

Reviewer: ```suggestion
We could add an argument to the `copy` key binding argument to allow for formatted copying when the user chooses to do so.
```

## UI/UX Design

### Settings option 1 - global setting

a. The user could list which kinds of formats they want included when they copy. When right clicking, they would copy with these formats. If this is set to something other than `plain`, holding `alt` and right clicking would copy plain text.

> `"copyFormats": "html,rtf,plain"`

Reviewer 2: Rrefer an array, because they're easier to parse (we don't have to handle `html,rtf,plain` and `html,          rtf, plain` and `rtf,plain,            html` and we don't need to error out on `html;rtf,plain` and stuff)

b. We could also just combine html and rtf into a single boolean. Users would either get plain text only (`false`) or all formatting (`true`) onto their clipboard. If this is set to `true`, the default right click behavior is reversed: right clicking copies the formatting and holding `alt` copies only the plain text.

`"copyFormatting": true`

### Settings option 2 - key binding argument

a. Just like the the `trimWhitespace` argument you can add to the `copy` key binding, we could add one for text formatting.

> `{"command": {"action": "copy", "keepFormatting": true}, "keys": "ctrl+a"}`

Reviewer: Personally, like this idea the best. It's simple and straightforward. At some point, after deciding which route to go with, we should decide what the default keybinding will be, if any.

> b. We could also split out the html and rtf formats.

Reviewer 2: call out downsides to making this an argument.

* no way to configure right-click

maybe note that option 1 and 2 are not mutually exclusive

mention something like that in "future considerations"

---
author: Kayla Cinnamon @cinnamon-msft
created on: <2020-04-01>
last updated: <2020-04-01>
issue id: <#4191>
---

# Formatted Copy

## Abstract

When copying text, the Terminal should provide the option of including formatting. Not all apps that receive text allow for picking which format you want when pasting. The default should be to only copy plain text, based on the response from this poll on Twitter.

![Twitter poll](twitter-poll.png)

## Solution Design

Solutions for the right click behavior and user settings are described below.

### Right click behavior

> Be default, right clicking to copy would only copy the plain text. If the user would like to copy the formatting, they can hold `alt` and right click (this behavior is the same in conhost).

Reviewer 3: ```suggestion
By default, right clicking to copy would only copy the plain text. If the user would like to copy the formatting, they can hold `alt` and right click (this behavior is the same in conhost).
```

### Settings option 1 - global setting

We could have a global setting that adds the formatting when the user copies. When set, this would change the right click behavior.

### Settings option 2 - key binding argument

We could add an argument to the `copy` key binding argument to allow for formatted copying when the user chooses to do so.

## UI/UX Design

### Settings option 1 - global setting

a. The user could list which kinds of formats they want included when they copy. When right clicking, they would copy with these formats. If this is set to something other than `plain`, holding `alt` and right clicking would copy plain text.

`"copyFormats": ["html","rtf","plain"]`

> b. We could also just combine html and rtf into a single boolean. Users would either get plain text only (`false`) or all formatting (`true`) onto their clipboard. If this is set to `true`, the default right click behavior is reversed: right clicking copies the formatting and holding `alt` copies only the plain text.

Reviewer 2: like other specs before it, make sure to call out the option we finally decided upon and move all unchosen options into an investigation notes section.

Reviewer 2: A spec isn't useful if it just says "we could do a, b, c" without it coming to its own consensus.

Author: I followed the same formatting as the [default profile settings spec](https://github.com/microsoft/terminal/blob/master/doc/specs/%232325%20-%20Default%20Profile%20Settings.md) and added a conclusions section at the bottom.

`"copyFormatting": true`

### Settings option 2 - key binding argument

> a. Just like the the `trimWhitespace` argument you can add to the `copy` key binding, we could add one for text formatting. This would not change the right click behavior.

Reviewer 3: ```suggestion
a. Just like the `trimWhitespace` argument you can add to the `copy` key binding, we could add one for text formatting. This would not change the right click behavior.
```

`{"command": {"action": "copy", "keepFormatting": true}, "keys": "ctrl+a"}`

b. We could also split out the html and rtf formats. The right click behavior would still stay as default.

`{"command": {"action": "copy", "formats": ["html","rtf","plain"]}, "keys": "ctrl+a"}`

## Capabilities

### Accessibility

This shouldn't affect accessibility.

### Security

This does not affect security.

### Reliability

This does not affect reliability.

### Compatibility

> This breaks the existing behavior of always copying the formatting. The justification for breaking this default behavior is in the response from the community saying the default should be plain text only.

Reviewer 3: ```suggestion
This breaks the existing behavior of always copying the formatting. The justification for breaking this default behavior is in response to the community saying the default should be plain text only.
```

### Performance, Power, and Efficiency

## Potential Issues

One possible issue is that discovering how to copy the formatting might be difficult to find. We could mitigate this by adding it into the settings.json file and commenting it out.

## Future considerations

> We could add the option we don't include later if people want it. For example, if we ship a global setting now, we could ship a key binding later that lets you choose how you want to copy, and vice versa.

Reviewer 3: The first sentence here doesn't parse. And I can't figure out how to fix it.

---
author: Kayla Cinnamon @cinnamon-msft
created on: 2020-04-01
last updated: 2020-04-07
issue id: #4191
---

# Formatted Copy

## Abstract

When copying text, the Terminal should provide the option of including formatting. Not all apps that receive text allow for picking which format you want when pasting. The default should be to only copy plain text, based on the response from this poll on Twitter.

![Twitter poll](twitter-poll.png)

## Solution Proposals

A proposal for the right click behavior as well as two user settings proposals are described below:

1. [Right click behavior](#right-click-behavior)
2. [Settings option 1 - global setting](#settings-option-1---global-setting)

> 3. [Settings option 2 - key binding argument](#settings-option-2---key-binding-argument)

Reviewer 4: This has got me all confused as I'm leaving comments on the doc - maybe make "Right Click Behavior" option _3_, so that "Settings option 1" is _1_, etc

### Right click behavior

By default, right clicking to copy would only copy the plain text. If the user would like to copy the formatting, they can hold `alt` and right click (this behavior is the same in conhost).

### Settings option 1 - global setting

> We could have a global setting that adds the formatting when the user copies. When set, this would change the right click behavior.

Reviewer 5: ```suggestion
We could have a global setting that when enabled, would copy formatting to the clipboard on all copy operations.
```
The second sentence seemed a little weird by only mentioning right click behavior, when it should be for all copies.

### Settings option 2 - key binding argument

We could add an argument to the `copy` key binding argument to allow for formatted copying when the user chooses to do so.

## UI/UX Design

### Settings option 1 - global setting

a. The user could list which kinds of formats they want included when they copy. When right clicking, they would copy with these formats. If this is set to something other than `plain`, holding `alt` and right clicking would copy plain text.

`"copyFormats": ["html","rtf","plain"]`

> b. We could also just combine html and rtf into a single boolean. Users would either get plain text only (`false`) or all formatting (`true`) onto their clipboard. If this is set to `true`, the default right click behavior is reversed: right clicking copies the formatting and holding `alt` copies only the plain text.

Reviewer 5: Correct me if I'm wrong, but if option b is the one we're going for, we're _not_ having the `Alt-RightClick` to copy plaintext/formatting right? Right click should just copy formatting if the setting is true, and plain text if false.

Author: Ah nice catch. I just removed references to `alt` right click functionality. 👍

`{"command": {"action": "copy", "keepFormatting": true}, "keys": "ctrl+a"}`

b. We could also split out the html and rtf formats. The right click behavior would still stay as default.

`{"command": {"action": "copy", "formats": ["html","rtf","plain"]}, "keys": "ctrl+a"}`

## Capabilities

### Accessibility

This shouldn't affect accessibility.

### Security

This does not affect security.

### Reliability

This does not affect reliability.

### Compatibility

This breaks the existing behavior of always copying the formatting. The justification for breaking this default behavior is in response to the community saying the default should be plain text only.

### Performance, Power, and Efficiency

## Potential Issues

One possible issue is that discovering how to copy the formatting might be difficult to find. We could mitigate this by adding it into the settings.json file and commenting it out.

## Future considerations

We could always add an additional option if people want more flexibility. For example, if we ship a global setting now, we could ship a key binding later that lets you choose how you want to copy, and vice versa.

## Conclusions

> The team has decided to have plain text as the default copy behavior and to enable formatted copying with a global setting that accepts a boolean value (settings option 1 - global setting, option b). In the future, we can modify this setting to also accept an array, so the user can specify which formats they would like to copy. Additionally, a key binding can be added to allow for greater flexibility.

Reviewer 4: Wait didn't we decide on a combo of both 2 & 3?

We do the global setting _now_, as a boolean, because that's easy. (This is 2a)

We can then update that to _additionally_ accept an array in the future, where `true` == `[html, rtf, plain]` and `false`==`[plain]` (This is 2b)

Then, we could further add an optional parameter to the copy `ShortcutAction`. If it's set, we'll copy with the provided formatting. Otherwise, we'll use the value of the global setting? (This is 3)

Reviewer 4: Okay I guess this is kinda what it _does_ say, I just didn't immediately parse that this was a phased implementation plan

---
author: Kayla Cinnamon @cinnamon-msft

> created on: <2020-04-01>

Reviewer 2: the reason these are in `<>` in the spec template is to indicate that the entire value is a placeholder and should be changed. The `<>` should not be retained in the actual spec document.
